cob_obstacle_distance/src/frame_to_collision.cpp

```diff
@@ -0,0 +1,300 @@
+    {
+        PtrCylinder_t urdf_cyl = boost::static_pointer_cast<urdf::Cylinder>(geometry);
+        fcl::Cylinder c(urdf_cyl->radius, urdf_cyl->length);
+        segment_of_interest_marker_shape.reset(new MarkerShape<fcl::Cylinder>(this->root_frame_id_, c, pose, col));
+    }
+    else
+    {
+        ROS_ERROR_STREAM("Geometry type unknown: " << geometry->type);
+    }
+}
+
+
+bool FrameToCollision::getMarkerShapeFromType(const uint32_t& shape_type,
+                                              const Eigen::Vector3d& abs_pos,
+                                              const Eigen::Quaterniond& quat_pos,
+                                              const std::string& frame_of_interest,
+                                              const Eigen::Vector3d& dimension,
+                                              PtrIMarkerShape_t& segment_of_interest_marker_shape)
+{
+    geometry_msgs::Pose pose;
+    tf::pointEigenToMsg(abs_pos, pose.position);
+    tf::quaternionEigenToMsg(quat_pos, pose.orientation);
+    return this->getMarkerShapeFromType(shape_type,
+                                        pose,
+                                        frame_of_interest,
+                                        dimension,
+                                        segment_of_interest_marker_shape);
+}
+
+
+bool FrameToCollision::getMarkerShapeFromType(const uint32_t& shape_type,
+                            const geometry_msgs::Pose& pose,
+                            const std::string& frame_of_interest,
+                            const Eigen::Vector3d& dimension,
+                            PtrIMarkerShape_t& segment_of_interest_marker_shape)
+{
+    // Representation of segment_of_interest as specific fcl::Shape
+    fcl::Box b(dimension(FCL_BOX_X), dimension(FCL_BOX_Y), dimension(FCL_BOX_Z));
+    fcl::Sphere s(dimension(FCL_RADIUS));
+    fcl::Cylinder c(dimension(FCL_RADIUS), dimension(FCL_CYL_LENGTH));
+
+    std_msgs::ColorRGBA col;
+    col.a = 1.0;
+    col.r = 1.0;
+    col.g = 0.0;
+    col.b = 0.0;
+
+    uint32_t loc_shape_type = shape_type;
+    std::string mesh_resource;
+    if(visualization_msgs::Marker::MESH_RESOURCE == loc_shape_type)
+    {
+        PtrConstLink_t link = this->model_.getLink(frame_of_interest);
+        if(this->success_ &&
+                NULL != link &&
+                NULL != link->collision &&
+                NULL != link->collision->geometry &&
+                link->collision->geometry->type == urdf::Geometry::MESH)
+        {
+            PtrMesh_t mesh = boost::static_pointer_cast<urdf::Mesh>(link->collision->geometry);
+            mesh_resource = mesh->filename;
+        }
+        else
+        {
+            ROS_WARN_STREAM("Either link is not available in URDF or there is no MESH collision representation for " <<
+                            frame_of_interest <<
+                            ". Rather using SPHERE.");
+            loc_shape_type = visualization_msgs::Marker::SPHERE;
```

Comment:
According to URDF [specification](http://wiki.ros.org/urdf/XML/link), it is required to specify a `geometry` in case you decided to specify a `collision` tag. And in `geometry`, a `box`, a `sphere` or a `cylinder` could be used as well.

Thus, should we consider box, sphere, cylinder here as well - if they are specified?
Should we use the `geometry`of the `visual` tag - in case no `collision` is specified but a `visual`?
I think it's a good idea to default to a `sphere` in case neither a `visual` nor a `collision` tag is specified, i.e. in "virtual links"!